syntax = "proto2";
package com.common.msg;

import "SceneModuleMsg.proto";
import "ItemModuleMsg.proto";

enum MagicalCreatureModuleMsgSubCommand {
	// 购买蛋
	MAGICALCREATUREMODULEMSG_SUB_BUYEGG = 24001;
	// 蛋合成魔法生物
	MAGICALCREATUREMODULEMSG_SUB_EGGSYNTHESIS = 24002;
	// 三个魔法生物合成升级
	MAGICALCREATUREMODULEMSG_SUB_CREATURELEVELUP = 24003;
	// 培育
	MAGICALCREATUREMODULEMSG_SUB_BREED = 24004;
	// 获取魔法生物数据
	MAGICALCREATUREMODULEMSG_SUB_GETCREATUREDATA = 24005;
	// 龙巢开启
	MAGICALCREATUREMODULEMSG_SUB_OPENNEST = 24006;
	// 龙巢升级
	MAGICALCREATUREMODULEMSG_SUB_NESTLEVELUP = 24007;
	// 领取指定龙的挂机奖励
	MAGICALCREATUREMODULEMSG_SUB_COLLECTHANGUPREWARD = 24008;
	// 领取所有挂机奖励
	MAGICALCREATUREMODULEMSG_SUB_COLLECTALLHANGUPREWARD = 24009;
	// 指定龙进入龙巢
	MAGICALCREATUREMODULEMSG_SUB_MOVETONEST = 24010;
	// 采集
	MAGICALCREATUREMODULEMSG_SUB_COLLECT = 24011;
	// 点击采集奖励
	MAGICALCREATUREMODULEMSG_SUB_COLLECTPLANTREWARD = 24012;
	// 加速挂机
	MAGICALCREATUREMODULEMSG_SUB_ACCELERATEHANGUP = 24013;
	// 喂食
	MAGICALCREATUREMODULEMSG_SUB_FEED = 24014;
	// 购买龙
	MAGICALCREATUREMODULEMSG_SUB_BUYCREATURE = 24015;
	// 抚摸龙奖励
	MAGICALCREATUREMODULEMSG_SUB_CARESSREWARD = 24016;
	// 宠物龙奖励
	MAGICALCREATUREMODULEMSG_SUB_PETREWARD = 24017;
	// 指引过程使龙进入饥饿状态
	MAGICALCREATUREMODULEMSG_SUB_GUIDEENTERHUNGER = 24018;
	// 补充龙体力
	MAGICALCREATUREMODULEMSG_SUB_BUYPHYSICALSTRENGTH = 24019;
	// 获取积攒的消耗体力累计值
	MAGICALCREATUREMODULEMSG_SUB_GETCONSUMEENERGY = 24020;
	// 选择探索龙
	MAGICALCREATUREMODULEMSG_SUB_SELECTEXPLORE = 24021;
	// 释放技能
	MAGICALCREATUREMODULEMSG_SUB_DOSKILL = 24022;
	// 开始委托
	MAGICALCREATUREMODULEMSG_SUB_COMMISSION = 24023;
	// 加速委托
	MAGICALCREATUREMODULEMSG_SUB_ACCELERATECOMMISSION = 24024;
	// 领取委托奖励
	MAGICALCREATUREMODULEMSG_SUB_COMMISSIONREWARD = 24025;
	// 解锁层
	MAGICALCREATUREMODULEMSG_SUB_COMMISSIONUNLOCK = 24026;
	// 出售龙
	MAGICALCREATUREMODULEMSG_SUB_SELLCREATURE = 24027;
	// 选择探索遗迹龙
	MAGICALCREATUREMODULEMSG_SUB_SELECTEXPLORERELIC = 24028;
	// 培育领取奖励
	MAGICALCREATUREMODULEMSG_SUB_BREEDREWARD = 24029;
	// 领取龙的挂机奖励
	MAGICALCREATUREMODULEMSG_SUB_CREATURESHANGUPREWARD = 24030;
	//请求皇冠抽奖
	MAGICALCREATUREMODULEMSG_SUB_CROWNDRAW = 24031;
	// 龙装扮信息
	MAGICALCREATUREMODULEMSG_SUB_CREATURESKINLIST = 24032;
	// 龙装扮装备请求
	MAGICALCREATUREMODULEMSG_SUB_CREATURESKINEQUIP = 24033;
	// 委托结束释放龙
	MAGICALCREATUREMODULEMSG_SUB_COMMISSIONFINISH = 24034;
	// 加速龙体力恢复
	MAGICALCREATUREMODULEMSG_SUB_ACCELERATERECOVERY = 24035;
	// 更换龙所在龙巢
	MAGICALCREATUREMODULEMSG_SUB_CHANGENEST = 24036;
	//请求离线喂龙数据(月卡新增功能)
	MAGICALCREATUREMODULEMSG_SUB_AUTOFEED = 24037;
	//请求在线喂龙
	MAGICALCREATUREMODULEMSG_SUB_ONLINEAUTOFEED = 24038;
}

message MagicalCreatureBuyMsg {
  // 龙或蛋或龙繁育模板id
  required string templateId = 1;
  // 购买次数
  required int32 buyCnt = 2;
  // 模板类型(0龙、1繁育)
  optional int32 type = 3;
  // 未出龙的繁育次数
  optional int32 notDragonBreedCnt = 4;
}

message MagicalNestMsg {
  // 巢类型
  required int32 nestType = 1;
  // 巢等级
  required int32 nestLevel = 2;
  // 巢额外增加上限（购买）
  optional int32 extraLimit = 3;
}

// 家园委托数据
message CommissionMsg {
	// 洞穴Id
	optional string commissionId = 1;
	// 开始时间
	optional int64 startTime = 2;
	// 是否完成 13.1之后不使用了
	optional bool finished = 3;
	// 龙列表
	repeated int64 creatures = 4;
	// 已解锁最大层洞穴Id 13.1之后不使用了
	optional string unlockMaxId = 5;
	// 龙模板id 释放龙后需要UI上显示
	repeated string creatureTemplateIds = 6;
}

// 购买蛋 请求
message BuyEggRequest {
	required string eggId = 1;
}

// 购买蛋 响应
message BuyEggResponse {
}

// 蛋合成魔法生物 请求
message EggSynthesisRequest {
	required string eggId = 1;
}

// 蛋合成魔法生物 响应
message EggSynthesisResponse {
	required MagicalCreatureMsg creature = 1;
}

// 三个魔法生物合成升级 请求
message CreatureLevelUpRequest {
	required int32 creatureId1 = 1;
	required int32 creatureId2 = 2;
	required int32 creatureId3 = 3;
}

// 三个魔法生物合成升级 响应
message CreatureLevelUpResponse {
	required MagicalCreatureMsg creature = 1;
	// 同步道具
	repeated ItemMsg items = 2;
	// 其他龙同步(龙收集系统影响其他龙之力)
	repeated MagicalCreatureMsg others = 3;
}

// 培育 请求
message BreedRequest {
	required string templateId1 = 1;
	required string templateId2 = 2;
	//活动触发的繁育需要传递对应活动的id
	optional string actId = 3;
	//繁育次数
	optional int32 count = 4;
	//是否使用了回流活动福利卡
	optional bool useWelfare = 5;
	// 看广告繁育
	optional bool ads = 6;
}

// 培育 响应
message BreedResponse {
	// 获得龙
	optional MagicalCreatureMsg creature = 1;
	// 同步道具
	repeated ItemMsg items = 2;
	// 培育产出道具
	repeated ItemMsg productItems = 3;
}

//魔法生物数据请求
message GetCreatureDataRequest {
}

//魔法生物数据响应
message GetCreatureDataResponse {
	//购买信息
	repeated MagicalCreatureBuyMsg buys = 1;
	//龙信息
	repeated MagicalCreatureMsg creatures = 2;
	//技能释放次数
	optional int32 skillCnt = 3;
	//技能释放次数刷新时间
	optional int64 refreshTime = 4;
	//培育Id
	optional string breedId = 5;
	//培育奖励次数
	optional int32 breedAwards = 6;
	//培育未得龙次数
	optional int32 notDragonBreedCnt = 7;
	//皇冠龙抽奖次数
	optional int32 crownDrawCnt = 8;
}

message OpenNestRequest {
	required int32 nestType = 1;
}

message OpenNestResponse {
}

message NestLevelUpRequest {
	required int32 nestType = 1;
}

message NestLevelUpResponse {
}

message CollectHangupRewardRequest {
	required int32 creatureId = 1;
	// 龙挂机收益物品id
	optional string itemId = 2;
	// 龙挂机收益物品数量 服务器已经弃用此字段
	optional int32 itemCnt = 3;
}

message CollectHangupRewardResponse {
	// 同步道具
	repeated ItemMsg items = 1;
	optional MagicalCreatureMsg creatures = 2;
	// 16.1 奖励倍数
	optional int32 rewardMulti = 3;
}

message CollectAllHangupRewardRequest {
}

message HangUpRewardMsg {
  // 龙id
  optional int64 creatureId = 1;
  // 产出Id
  optional string itemId = 2;
  // 产出Id
  optional int32 itemCnt = 3;
  // 奖励倍数
  optional int32 rewardMulti = 4;
}

message CollectAllHangupRewardResponse {
	// 同步道具
	repeated ItemMsg items = 1;
	repeated MagicalCreatureMsg creatures = 2;
	//道具数量变化
	repeated int32 itemChangeNum = 3;
	// 16.1 奖励倍数信息 只在有倍数情况下同步
	repeated HangUpRewardMsg multiRewards = 4;
}

message MoveToNestRequest {
	required int32 creatureId = 1;
	required int32 nestType = 2;
}

message MoveToNestResponse {
	optional MagicalCreatureMsg creatures = 1;
}

// 采集请求
message CollectRequest {
	required int32 creatureId = 1;
	// 场景ID
	optional string sceneId = 2;
	// 障碍物ID
	optional string plantId = 3;
}

// 采集响应
message CollectResponse {
	required MagicalCreatureMsg creature = 1;
	optional PlantMsg plant = 2;
	// version>=4.1 收集障碍物成就奖励
	optional ObstacleAchievementRewardMsg achievementReward = 3;
	// version>=16.1 龙pve场景伤害暴击buff倍数
	optional int32 damageCrit = 4;
	// version>=16.1 龙pve场景掉落暴击buff倍数
	optional int32 dropCrit = 5;
	// version>=16.1 采集时装扮额外获得产出
	repeated ItemMsg skinRewards = 6;
}

// 领取障碍物龙采集奖励请求
message CollectPlantRewardRequest {
	// 场景ID
	optional string sceneId = 1;
	// 障碍物ID
	optional string plantId = 2;
}

// 领取障碍物龙采集奖励返回
message CollectPlantRewardResponse {
	// 采集障碍物结果
	optional ClearPlantResponse clearRet = 1;
}

// 加速挂机请求
message AccelerateHangUpRequest {
	required int32 creatureId = 1;
	// 消耗item
	required ItemMsg consume = 2;
}

// 加速挂机返回
message AccelerateHangUpResponse {
	// 同步道具
	repeated ItemMsg items = 1;
	optional MagicalCreatureMsg creatures = 2;
}

message FeedMsg {
	required int32 creatureId = 1;
	required string itemId = 2;
	required int64 feedTime = 3;
}

// 喂食请求
message FeedRequest {
	repeated FeedMsg feedMsgs = 1;
}

// 喂食返回
message FeedResponse {
	// 同步道具
	repeated ItemMsg items = 1;
	repeated MagicalCreatureMsg creatures = 2;
}

// 购买龙请求
message BuyCreatureRequest {
	required string templateId = 1;
	// 购买途径 1合成界面，2龙获取界面
	optional int32 way = 2;
}

// 购买龙返回
message BuyCreatureResponse {
	required MagicalCreatureMsg creature = 1;
}

// 抚摸龙奖励请求
message CaressRewardRequest {
	required int32 creatureId = 1;
	required bool reset = 2;
}

// 抚摸龙奖励返回
message CaressRewardResponse {
	required string rewardItemId = 1;
	required int32 rewardCount = 2;
}

// 宠物龙奖励请求
message PetRewardRequest {
}

// 宠物龙奖励返回
message PetRewardResponse {
	// 奖励道具
	required ItemMsg rewardItem = 1;
}

message GuideEnterHungerRequest {
	required int32 creatureId = 1;
}

message GuideEnterHungerResponse {
}

message BuyPhysicalStrengthRequest {
	required int32 creatureId = 1;
	required int32 price = 2;
}

message BuyPhysicalStrengthResponse {
	required MagicalCreatureMsg creature = 1;
	// 同步道具
	repeated ItemMsg items = 2;	
}

message GetConsumeEnergyRequest {
}

message GetConsumeEnergyResponse {
	required int32 consumeEnergyCnt = 1;
}

message SelectExploreRequest {
	repeated int32 creatureIds = 1;
}

message SelectExploreResponse {
	// 取消探索的龙会重新计算挂机开始时间
	repeated MagicalCreatureMsg creatures = 1;
}

message DoSkillRequest {
	// 龙ID
	required int32 creatureId = 1;
	// 释放技能ID
	optional string skillId = 2;
	// 场景ID
	optional string sceneId = 3;
	// 释放位置x
	optional int32 x = 4;
	// 释放位置y
	optional int32 y = 5;
	// 目标
	repeated string plantIds = 6;
}

message DoSkillResponse {
	// 技能清除障碍效果
	repeated ClearPlantResponse effects = 1;
	// 技能奖励道具
	repeated ItemMsg items = 2;	
}

// 委托请求
message CommissionRequest {
	// 洞穴Id
	optional string commissionId = 1;
	// 选择龙列表
	repeated int64 creatureIds = 2;
}

// 委托响应
message CommissionResponse {
	// 道具同步
	repeated ItemMsg items = 1;
	// 龙信息同步
	repeated MagicalCreatureMsg creatures = 2;
}

// 加速委托请求
message AccelerateCommissionRequest {
	// 洞穴Id
	optional string commissionId = 1;
	// 道具id
	optional string itemId = 2;
	// 看广告免费加速
	optional bool ads = 3;
}

// 加速委托响应
message AccelerateCommissionResponse {
	// 道具同步
	repeated ItemMsg items = 1;
	// 加速结果
	optional CommissionMsg commission = 2;
}

// 领取委托奖励请求
message CommissionRewardRequest {
	// 洞穴Id
	optional string commissionId = 1;
}

// 领取委托奖励响应
message CommissionRewardResponse {
	// 龙信息同步
	repeated MagicalCreatureMsg creatures = 1;
	// 奖励道具同步
	repeated ItemMsg items = 2;
}

// 解锁层请求
message CommissionUnlockRequest {
	// 洞穴Id
	optional string commissionId = 1;
}

// 解锁层响应
message CommissionUnlockResponse {
}

// 出售龙请求
message SellCreatureRequest {
	// 龙id
	optional int64 creatureId = 2;
}

// 出售龙返回
message SellCreatureResponse {
	repeated ItemMsg items = 1;
	// 其他龙同步(龙收集系统影响其他龙之力)
	repeated MagicalCreatureMsg others = 2;
}

message SelectExploreRelicRequest {
	optional string sceneId = 1;
	repeated int32 creatureIds = 2;
}

message SelectExploreRelicResponse {
	// 取消探索的龙会重新计算挂机开始时间
	repeated MagicalCreatureMsg creatures = 1;
}

// 培育领取奖励 请求
message BreedRewardRequest {
}

// 培育领取奖励 响应
message BreedRewardResponse {
	// 获得龙
	optional MagicalCreatureMsg creature = 1;
	// 同步道具
	repeated ItemMsg items = 2;
	// 培育产出道具
	repeated ItemMsg productItems = 3;
	//获取的多个龙的其他龙数据同步
	repeated MagicalCreatureMsg others = 4;
	// since 22.1 龙繁育次数变化同步
	repeated MagicalCreatureBuyMsg buys = 5;
}

// 领取龙的挂机奖励 请求
message CreaturesHangupRewardRequest {
	repeated int32 creatureIds = 1;
}

// 领取龙的挂机奖励 响应
message CreaturesHangupRewardResponse {
	// 同步道具
	repeated ItemMsg items = 1;
	// 同步龙
	repeated MagicalCreatureMsg creatures = 2;
	//道具数量变化
	repeated int32 itemChangeNum = 3;
}

// 皇冠抽奖请求
message CrownDrawRequest {
}

// 皇冠抽奖请求 响应
message CrownDrawResponse {
	//奖励的道具
	optional ItemMsg item = 1;
}

// 龙装扮列表请求
message CreatureSkinListRequest {
	
}

// 龙装扮列表响应
message CreatureSkinListResponse {
	// 装扮列表
	repeated DragonSkinMsg skins = 1;
}

// 龙装扮装备请求
message CreatureSkinEquipRequest {
	// 龙ID
	optional int64 creatureId = 1;
	// 装扮实例ID
	optional int64 skinId = 2;
	// 装备类型 1:装备 2:卸下
	optional int32 op = 3;
}

// 龙装扮装备响应
message CreatureSkinEquipResponse {
	// 龙信息同步
	optional MagicalCreatureMsg creature = 1;
}

// 委托结束释放龙请求
message CommissionFinishRequest {
	// 洞穴Id
	optional string commissionId = 1;
}

// 委托结束释放龙响应
message CommissionFinishResponse {
	// 龙信息同步
	repeated MagicalCreatureMsg creatures = 1;
}

// 加速龙体力恢复请求
message AccelerateRecoveryRequest {
	required int32 creatureId = 1;
	required string itemId = 2;
	required int32 count = 3;
	// 看广告加速恢复
	optional bool ads = 4;
}

// 加速龙体力恢复响应
message AccelerateRecoveryResponse {
	required MagicalCreatureMsg creature = 1;
	// 同步道具
	repeated ItemMsg items = 2;	
}

// 更换龙所在龙巢信息
message ChangeNestMsg {
  // 巢类型
  required int64 creatureId = 1;
  // 巢类型
  required int32 nestType = 2;
}

// 更换龙所在龙巢请求
message ChangeNestRequest {
	repeated ChangeNestMsg changeNests = 1;
	
}

// 更换龙所在龙巢响应
message ChangeNestResponse {
	repeated MagicalCreatureMsg creatures = 1;
}

// 请求离线喂龙数据(月卡新增功能)
message AutoFeedRequest {
}

message AutoFeedRet {
	//所有自动喂完龙的数据同步
	repeated MagicalCreatureMsg creatures = 1;
	//消耗道具
	repeated ItemMsg costItems = 2;
	//拆分后新增的道具
	repeated ItemMsg addItems = 3;
	//是否有道具不足的情况
	optional bool itemNotEnough = 4;
}

// 自动喂龙数据(月卡新增功能)响应
message AutoFeedResponse {
	optional AutoFeedRet autoFeedRet = 1;
}

//请求在线自动喂龙
message OnlineAutoFeedRequest {
	//请求要喂的龙的集合
	repeated int64 creatureIds = 1;
}

//请求在线自动喂龙响应
message OnlineAutoFeedResponse {
	optional AutoFeedRet autoFeedRet = 1;
	//没有喂成功的龙的集合
	repeated MagicalCreatureMsg notFeedCreatures = 2;
}